rosidl_runtime_c/src/type_hash.c

```diff
@@ -0,0 +1,151 @@
+  const rosidl_type_hash_t * type_hash,
+  rcutils_allocator_t allocator,
+  char ** output_string)
+{
+  RCUTILS_CHECK_ARGUMENT_FOR_NULL(type_hash, RCUTILS_RET_INVALID_ARGUMENT);
+  if (!rcutils_allocator_is_valid(&allocator)) {
+    RCUTILS_SET_ERROR_MSG("Invalid allocator");
+    return RCUTILS_RET_INVALID_ARGUMENT;
+  }
+  RCUTILS_CHECK_ARGUMENT_FOR_NULL(output_string, RCUTILS_RET_INVALID_ARGUMENT);
+
+  char * local_output = allocator.allocate(RIHS01_STRING_LEN + 1, allocator.state);
+  if (!local_output) {
+    *output_string = NULL;
+    RCUTILS_SET_ERROR_MSG("Unable to allocate space for type hash string.");
+    return RCUTILS_RET_BAD_ALLOC;
+  }
+  local_output[RIHS01_STRING_LEN] = '\0';
+  memcpy(local_output, RIHS01_PREFIX, RIHS_PREFIX_LEN);
+  for (size_t i = 0; i < ROSIDL_TYPE_HASH_SIZE; i++) {
+    _xitoa(type_hash->value[i], local_output + RIHS_PREFIX_LEN + (i * 2));
+  }
+
+  *output_string = local_output;
+  return RCUTILS_RET_OK;
+}
+
+rcutils_ret_t
+rosidl_parse_type_hash_string(
+  const char * type_hash_string,
+  rosidl_type_hash_t * hash_out)
+{
+  RCUTILS_CHECK_ARGUMENT_FOR_NULL(type_hash_string, RCUTILS_RET_INVALID_ARGUMENT);
+  RCUTILS_CHECK_ARGUMENT_FOR_NULL(hash_out, RCUTILS_RET_INVALID_ARGUMENT);
+  hash_out->version = 0;
+  size_t input_len = strlen(type_hash_string);
+
+  // Check prefix
+  if (input_len < RIHS_PREFIX_LEN) {
+    RCUTILS_SET_ERROR_MSG("Hash string not long enough to contain RIHS prefix.");
+    return RCUTILS_RET_INVALID_ARGUMENT;
+  }
+  if (0 != strncmp(type_hash_string, RIHS01_PREFIX, RIHS_VERSION_IDX)) {
+    RCUTILS_SET_ERROR_MSG("Hash string doesn't start with RIHS.");
+    return RCUTILS_RET_INVALID_ARGUMENT;
+  }
+
+  // Parse version
+  char version_top = type_hash_string[RIHS_VERSION_IDX];
+  char version_bot = type_hash_string[RIHS_VERSION_IDX + 1];
+  if (!(_ishexdigit(version_top) && _ishexdigit(version_bot))) {
+    RCUTILS_SET_ERROR_MSG("RIHS version is not a 2-digit hex string.");
+    return RCUTILS_RET_INVALID_ARGUMENT;
+  }
+  hash_out->version = (_xatoi(version_top) << 4) + _xatoi(version_bot);
+
+  if (hash_out->version != 1) {
+    RCUTILS_SET_ERROR_MSG("Do not know how to parse RIHS version.");
+    return RCUTILS_RET_INVALID_ARGUMENT;
+  }
+  if (input_len != RIHS01_STRING_LEN) {
+    RCUTILS_SET_ERROR_MSG("RIHS string is the incorrect size to contain a RIHS01 value.");
+    return RCUTILS_RET_INVALID_ARGUMENT;
+  }
+  const char * value_str = type_hash_string + RIHS_PREFIX_LEN;
+  for (size_t i = 0; i < ROSIDL_TYPE_HASH_SIZE * 2; i++) {
+    if (!_ishexdigit(value_str[i])) {
+      RCUTILS_SET_ERROR_MSG("Type hash string value contained non-hex-digit character.");
+      return RCUTILS_RET_INVALID_ARGUMENT;
+    }
+  }
+  for (size_t i = 0; i < ROSIDL_TYPE_HASH_SIZE; i++) {
+    // No error checking on byte values because the whole string was checked in prior loop
+    uint8_t byte_val = (_xatoi(value_str[i * 2]) << 4) + _xatoi(value_str[(i * 2) + 1]);
```

Comment:
Same problem with the `xatoi` here.

Reply:
I left the above comment for this reason - we've already done the error checking on this value, `_xatoi` cannot return -1 in this loop (unless some multithreading intentional crime modifying the value as we're translating it?). This is the case for the version above too, the error checking is done before translation, not during it. I wanted to keep those helper functions as hidden static ones in this unit for the purpose of not exposing that specific usage to the outside.

At this point, it's `if-above` or `if-within`, the effect is the same.

Reply:
Sorry for the pushback, it's part of my problem solving process. Though I could do it internally. I think the code looks a fine mostly inlined with error checking in-place, so I've done a pass that looks that way and will mark these 3 comments resolved, please reopen with any new comments.